### PR TITLE
BLD: Use cibuildwheel to build and deploy to PyPI

### PR DESCRIPTION
**Pull request by bocklund**
## Changes

- Use PyPA/cibuildwheel to build wheels (and source) distributions and deploy them to PyPI.
- The builds are run and artifacts are uploaded to GitHub on every push and pull request. The artifacts are only deployed to PyPI if the workflow is triggered by a GitHub Release event.
- Drops Python 3.6 following (NEP 29 recommendation)
- Pin to pyparsing v2 because v3 seemed to result in some breakages (it's still in pre-release stage, I did not investigate the error)
- Update RELEASING document with the new steps for creating a release to trigger CI. Make the steps more bulletproof. Build the document into the new "Developer" section on the docs website.

## Try out the package

I made some test alpha releases on test.pypi.org (most up-to-date is 0.9.0a1.post2). *This is just on test.pypi.org. We can wipe it clean after merging this PR*. The pre-releases can be installed with

* `pip install --pre --extra-index-url https://test.pypi.org/simple "pycalphad>0.8.5"`
* `pytest --pyargs pycalphad`

## Testing the artifacts in CI
Because the cibuildwheel process is already long running relative to our platform build matrix tests, I opted to not use the [cibuildwheel testing feature](https://cibuildwheel.readthedocs.io/en/stable/options/#testing). Since releasing is so rare and we want to be sure we get it right, I'm open to trying it out.

## Using the GitHub uploaded artifacts

Since the build artifacts are uploaded to GitHub, we could manually pull them down for local testing or have a job step to put them on an S3 bucket at some point. Not implemented here at this time.

### CODE REVIEW

**Comment by bocklund** [comment]
The 0.9.0a1.post1 build is still in process: https://github.com/bocklund/pycalphad/actions/runs/903644969

0.9.0a1 works, but pulls down a pre-release of pyparsing v3. Installing our pre-release of 0.9.0a1, then uninstalling pyparsing v3 in favor of a v2 gave working tests.

**Comment by bocklund** [comment]
I made some changes to the RELEASING doc that outlines the new release and deployment steps. I also included the steps in a new "Developer" toctree in the docs, so it will be on the website now. Hopefully it'll be a lot easier to follow and copy-paste steps this way.

## Screenshots of the changes to the docs:

### New sidebar while on the main page


Screen Shot 2021-06-03 at 3 38 13 PM

### Releasing page with expanded sidebar

Screen Shot 2021-06-03 at 3 38 01 PM

**Comment by bocklund** [comment]
For reference, the archive can be downloaded from 
Screen Shot 2021-06-03 at 4 34 05 PM

File: .github/workflows/deploy.yaml [hunk 1]
```diff
@@ -1,27 +1,81 @@
...
       - uses: actions/setup-python@v2
+        name: Install Python
         with:
-          python-version: '3.x'
-      - run: pip install twine build
-      - run: python -m build --sdist
-      - run: twine check --strict dist/*
-      - name: "Upload to PyPI"
-        run: twine upload -r pypi dist/pycalphad-*.tar.gz
+          python-version: '3.8'
```
Comment (richardotis) [inline .github/workflows/deploy.yaml:29]: Does cibuildwheel use its own Python for building each version? Will py37 and py39 wheels still be built even though py38 is specified here?

File: .github/workflows/deploy.yaml [hunk 2]
```diff
@@ -1,27 +1,81 @@
...
+
+      - uses: actions/upload-artifact@v2
+        with:
+          path: dist/*.tar.gz
+
+  upload_pypi:
+    needs: [build_wheels, build_sdist]
+    runs-on: ubuntu-latest
+    # upload to PyPI when a GitHub Release is created
+    if: github.event_name == 'release' && github.event.action == 'published'
```
Comment (richardotis) [inline .github/workflows/deploy.yaml:68]: So wheels are built on every commit, and only deployed when a GitHub Release is created?

File: RELEASING.rst [hunk 3]
```diff
@@ -1,47 +1,63 @@
...
-   Assuming a correctly configured .pypirc:
+#. ``python -m build --sdist``
+#. **Make sure that the script correctly detected the new version exactly and not a dirty / revised state of the repo.**
+#. ``twine upload dist/*`` to upload (assumes a `correctly configured <https://packaging.python.org/specifications/pypirc/>`_ ``~/.pypirc`` file)
 
-   ``twine upload -r pypi -u rotis dist/*``
 
-Uploading to conda-forge
-------------------------
+Deploy to conda-forge
```
Comment (richardotis) [inline RELEASING.rst:59]: Will this manual step still be necessary for conda-forge? Will auto PRs be opened for our PyPI source distributions?

File: environment-dev.yml [hunk 4]
```diff
@@ -14,8 +14,8 @@ dependencies:
   - importlib_metadata  # drop when pycalphad drops support for Python<3.8
   - matplotlib-base>=3.3
   - numpy>=1.13
-  - pyparsing
-  - python>=3.6
+  - pyparsing>=2,<3
```
Comment (richardotis) [inline environment-dev.yml:17]: This is fine, and it's not a blocker for 0.9, but please open an issue to track pyparsing 3 errors.

File: .github/workflows/deploy.yaml (hunk 2 above)
Comment (bocklund) [reply .github/workflows/deploy.yaml:68]: Correct

File: .github/workflows/deploy.yaml (hunk 1 above)
Comment (bocklund) [reply .github/workflows/deploy.yaml:29]: > Does cibuildwheel use its own Python for building each version?

Correct.

> Will py37 and py39 wheels still be built even though py38 is specified here?

Yes, the py38 is what is used to run cibuildwheel itself. Internally, cibuildwheel has it's own build matrix of platforms x Python versions where it does isolated installs of the different Python versions and builds the wheels accordingly.

File: RELEASING.rst (hunk 3 above)
Comment (bocklund) [reply RELEASING.rst:59]: PRs will continue to open automatically. This section is out of date and can probably be moved to a similar "manual" heading as PyPI.

I was honestly just too tired of writing all the main releasing and PyPI steps to have the energy to get to this one.

File: environment-dev.yml (hunk 4 above)
Comment (bocklund) [reply environment-dev.yml:17]: Done in #347

Comment (richardotis) [reply environment-dev.yml:17]: @bocklund Thanks for opening #347

**Review by richardotis — APPROVED** [review]

**Pull request merged**